x/mongo/driver/operation.go

```diff
@@ -1260,6 +1250,35 @@ func (op Operation) addClusterTime(dst []byte, desc description.SelectedServer)
 	// return bsoncore.AppendDocumentElement(dst, "$clusterTime", clusterTime)
 }
 
+// calculateMaxTimeMS calculates the value of the 'maxTimeMS' field to potentially append
+// to the wire message based on the current context's deadline and the 90th percentile RTT
+// if the ctx is a Timeout context. If the context is not a Timeout context, it uses the
+// operation's MaxTimeMS if set. If no MaxTimeMS is set on the operation, and context is
+// not a Timeout context, calculateMaxTimeMS returns 0.
+func (op Operation) calculateMaxTimeMS(ctx context.Context, rtt90 time.Duration) (uint64, error) {
+	if internal.IsTimeoutContext(ctx) {
+		if deadline, ok := ctx.Deadline(); ok {
+			remainingTimeout := time.Until(deadline)
+			// We check that remainingTimeout <= rtt90 because in the rare case that they are equal, we
+			// may accidentally calculate a maxTimeMS value of 0, which would indicate no timeout
+			// server-side.
+			if remainingTimeout <= rtt90 {
+				return 0, internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
+					"remaining time %v until context deadline is less than or equal to 90th percentile RTT", time.Until(deadline))
+			}
+			return uint64(remainingTimeout/time.Millisecond - rtt90/time.Millisecond), nil
+		}
+	} else if op.MaxTime != nil {
+		// Users are not allowed to pass a negative value as MaxTime. A value of 0 would indicate
+		// no timeout and is allowed.
+		if *op.MaxTime < 0 {
+			return 0, ErrNegativeMaxTime
+		}
+		return uint64(*op.MaxTime / time.Millisecond), nil
```

Comment:
It's possible for `calculateMaxTimeMS` to return 0 even if `op.MaxTime > 0` when the value is less than 1ms. Consider the following example:
```go
duration := 400 * time.Microsecond
op.MaxTime = &duration
fmt.Println(*op.MaxTime > 0) // Prints "true"
fmt.Println(uint64(*op.MaxTime / time.Millisecond)) // Prints "0"
```

Update to round the value up to the next millisecond value.

E.g.
```go
// Always round up to the next millisecond value so we never truncate the requested
// MaxTime value (e.g. 400 microseconds evaluates to 1ms, not 0ms).
return uint64((*op.MaxTime + (time.Millisecond - 1)) / time.Millisecond), nil
```

Reply:
Added.